substrate/client/consensus/beefy/src/lib.rs

```diff
@@ -292,21 +295,35 @@ pub async fn start_beefy_gadget<B, BE, C, N, P, R, S>(
 	// select recoverable errors.
 	loop {
 		// Wait for BEEFY pallet to be active before starting voter.
-		let persisted_state = match wait_for_runtime_pallet(
+		let (beefy_genesis, best_grandpa) = match wait_for_runtime_pallet(
 			&*runtime,
 			&mut beefy_comms.gossip_engine,
 			&mut finality_notifications,
 		)
 		.await
-		.and_then(|(beefy_genesis, best_grandpa)| {
-			load_or_init_voter_state(
-				&*backend,
-				&*runtime,
-				beefy_genesis,
-				best_grandpa,
-				min_block_delta,
-			)
-		}) {
+		{
+			Ok(res) => res,
+			Err(e) => {
+				error!(target: LOG_TARGET, "Error: {:?}. Terminating.", e);
+				return
+			},
+		};
+
+		if let Err(e) =
+			wait_for_headers_sync(&*backend, beefy_genesis, best_grandpa.clone(), HEADER_SYNC_DELAY)
+				.await
+		{
+			error!(target: LOG_TARGET, "Error: {:?}. Terminating.", e);
+			return
+		}
```

Comment:
This is expensive (especially when the chain is long), so we should do it _only_ when necessary. Most of the time BEEFY gadget is initialized from aux db from some existing state, in which case it had already handled `[beefy_genesis, best_beefy]` sub-chain.

You should move this logic after loading persisted state, but before finishing worker initialization and set the start of the range to `max(beefy_genesis, best_beefy)`.

Reply:
Good point. Changed the strategy to calling `wait_for_parent_header()` directly from `initialize_voter_state()`